filebeat/input/v2/input-cursor/clean.go

```diff
@@ -41,5 +43,81 @@ type cleaner struct {
 // for a long time, and the life time has been exhausted, then the resource will be removed immediately
 // once the last event has been ACKed.
 func (c *cleaner) run(canceler unison.Canceler, store *store, interval time.Duration) {
-	panic("TODO: implement me")
+	started := time.Now()
+	timed.Periodic(canceler, interval, func() {
+		gcStore(c.log, started, store)
+	})
+}
+
+// gcStore looks for resources to remove and deletes these. `gcStore` receives
+// the start timestamp of the cleaner as reference. If we have entries without
+// updates in the registry, that are older than `started`, we will use `started
+// + ttl` to decide if an entry will be removed. This way old entries are not
+// removed immediately on startup if the Beat is down for a longer period of
+// time.
+func gcStore(log *logp.Logger, started time.Time, store *store) {
+	log.Debugf("Start store cleanup")
+	defer log.Debugf("Done store cleanup")
+
+	states := store.ephemeralStore
+	states.mu.Lock()
+	defer states.mu.Unlock()
+
+	keys := gcFind(states.table, started, time.Now())
+	if len(keys) == 0 {
+		log.Debug("No entries to remove found")
+		return
+	}
+
+	if err := gcClean(store, keys); err != nil {
+		log.Errorf("Failed to remove all entries from the registry: %+v", err)
+	}
+}
+
+// gcFind searches the store of resources that can be removed. A set of keys to delete is returned.
+func gcFind(table map[string]*resource, started, now time.Time) map[string]struct{} {
+	keys := map[string]struct{}{}
+	for key, resource := range table {
+		clean := checkCleanResource(started, now, resource)
+		if !clean {
+			// do not clean the resource if it is still live or not serialized to the persistent store yet.
+			continue
+		}
+		keys[key] = struct{}{}
```

Comment:
Why are the values deleted in this function? To me, it seems this call belongs rather to `gcClean` function, as that is responsible for removing key-value pairs from the store.

Reply:
They are not deleted here. Cleanup happens in 2 phases: collect states/keys to delete -> remove entries from the k/v store. `gcFind` checks all states if they can be deleted. If `checkCleanResource` returns true we add the key to the set of states that we can delete and finally return the set. `gcClean` will do the final cleanup.